underfs/gsg/src/main/java/alluxio/underfs/gsg/GSGInputStream.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs.gsg;
+
+import com.google.cloud.ReadChannel;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A stream for reading a file from GSG. The main purpose is to provide a faster skip method, as
+ * the underlying implementation will read and discard bytes until the number to skip has been
+ * reached. This input stream returns 0 when calling read with an empty buffer.
+ */
+@NotThreadSafe
+public final class GSGInputStream extends InputStream {
+  /** Bucket name of the Alluxio GSG bucket. */
+  private final String mBucketName;
+
+  /** Key of the file in GSG to read. */
+  private final String mKey;
+
+  /** The Google cloud storage client. */
+  private final Storage mClient;
+
+  /** The pre-allocated buffer for single byte read operation. */
+  private final ByteBuffer mSingleByteBuffer;
+
+  /** The underlying input stream. */
+  private ReadChannel mReadChannel;
+
+  /** Position of the stream. */
+  private long mPos;
+
+  /**
+   * Creates a new instance of {@link GSGInputStream}, at a specific position.
+   *
+   * @param bucketName the name of the bucket
+   * @param key the key of the file
+   * @param client the client for GSG
+   * @param pos the position to start
+   */
+  GSGInputStream(String bucketName, String key, Storage client, long pos) {
+    mBucketName = bucketName;
+    mKey = key;
+    mClient = client;
+    mPos = pos;
+    mSingleByteBuffer = ByteBuffer.allocate(1);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mReadChannel != null) {
+      mReadChannel.close();
+    }
+  }
+
+  @Override
+  public int read() throws IOException {
+    if (mReadChannel == null) {
+      openStream();
+    }
+    mSingleByteBuffer.clear();
+    int num = mReadChannel.read(mSingleByteBuffer);
+    if (num != -1) { // valid data read
+      mPos++;
+    }
```

Comment:
Do we need to return `-1` here if `num == -1`?

Reply:
Changed to -1! Thanks for pointing out.